nm/checkpoint: Remove dependency on python-dbus

cathay4t opened the pull request:
nm/checkpoint: Remove dependency on python-dbus

Replacing current dbus checkpoint code with these libnm functions:
 * `libnm.NM.nmclient.checkpoint_create()`.
 * `libnm.NM.nmclient.checkpoint_rollback()`.
 * `libnm.NM.nmclient.checkpoint_destroy()`.

These functions only exists on libnm 1.12+(RHEL/CentOS 7.6 are shipping
libnm 1.12).
If libnm too old, raise `LibNMNotSupportCheckPoint` exception.

cathay4t commented:
@tyll Do you know whether `mainloop.run()` to wait on a action is best practise or not?

If possible, could you try apply a false config to trigger rollback?
I got below warning and has no idea why:
```
/usr/lib64/python3.7/site-packages/gi/overrides/GLib.py:498: Warning: g_atomic_ref_count_dec: assertion 'g_atomic_int_get (arc) > 0' failed
```

thom311 commented:
> > @tyll Do you know whether mainloop.run() to wait on a action is best practise or not?

time.sleep() is certainly not good, because the potential events on the mainloop (e.g. D-Bus events) are not processed and queue up. `g_mainloop_run()` is fine. Of course, then you may be called back from the mainloop while you are "sleeping", and the state after returning from the sleep, may have changed significantly. E.g. if you had a reference to an `NMDevice` which was previously tracked by `NMClient`, that device might have been deleted from `NMClient`'s cache (but the reference you hold, is of course still valid). Basically, it means you need to re-check the state that you care about to decide how to proceed.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            logging.warning('No checkpoint to rollback')
+
+def _checkpoint_create_callback(client, result, data):
+    mainloop = data._mainloop
```
Perhaps use `checkpoint = data` to clarify what is data and then calling its attributes will be clearer.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+    try:
+        cp = client.checkpoint_create_finish(result)
+        data._checkpoint = cp
+        if cp is None:
```
Can we reverse the other? It is strange to see the error handled first.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            mainloop.execute_next_action()
+    except Exception as e:
+        if mainloop.is_action_canceled(e):
+            logging.debug('Checkpoint creating canceled: error=%s', e)
```
`creation`

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+        if mainloop.is_action_canceled(e):
+            logging.debug('Checkpoint creating canceled: error=%s', e)
+        else:
+            mainloop.quit('Checkpoint creating failed: error={}'.format(e))
```
`creation`

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            logging.debug('Checkpoint creating canceled: error=%s', e)
+        else:
+            mainloop.quit('Checkpoint creating failed: error={}'.format(e))
+        return
```
Redundant

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            userdata = self
+            self._mainloop.push_action(
+                self._client.checkpoint_destroy,
+                self._checkpoint.get_path(), None,
```
One per line please and place the `None` in a descriptive variable so we will know what it represents.

cathay4t replied:
Done

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            userdata = self
+            self._mainloop.push_action(
+                self._client.checkpoint_rollback,
+                self._checkpoint.get_path(), None,
```
One per line please and place the `None` in a descriptive variable so we will know what it represents.

cathay4t replied:
I changed the `None` to `self._mainloop.cancellable`.

And done for the one per line.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                self._checkpoint.get_path(), None,
+                _checkpoint_rollback_callback, userdata)
+            self._mainloop.run(self._timeout)
+        else:
```
This scenario should not ever happen. If there is no cp, then we should have failed with an exception and never got here. Do you see any other case in which this is reasonable possible?

cathay4t replied:
I normally code things for worst case.

When some dev call rollback without create first, without any warning, he/she might take hours to figure out the cause.

EdDev replied:
IMO it should raise. (otherwise, it is a quiet error)

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+    try:
+        ret = client.checkpoint_rollback_finish(result)
+        logging.debug('Checkpoint %s rollback executed', cp.get_path())
+        for path in ret:
```
Please extract this for/block to a func.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -14,68 +14,25 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 #
+import time
```
Not used.

cathay4t replied:
Done.

cathay4t commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            devs, timeout, cp_flags, None, _checkpoint_create_callback,
+            userdata)
+
+        self._mainloop.run(timeout)
```
Done.

cathay4t commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +49,97 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+
+        self._mainloop.push_action(
+            self._client.checkpoint_create,
+            devs, timeout, cp_flags, None, _checkpoint_create_callback,
```
Done.

cathay4t commented on libnmstate/nm/checkpoint.py:
```diff
@@ -14,68 +14,25 @@
-        self._manager = nmdbus_manager()
         self._timeout = timeout
         self._checkpoint = None
+        self._client = nmclient.client()
```
Done.

cathay4t commented on libnmstate/nm/checkpoint.py:
```diff
@@ -14,68 +14,25 @@
-                                           _NMDbusManager.OBJ_PATH)
-        self.properties = dbus.Interface(mng_proxy, DBUS_STD_PROPERTIES_IFNAME)
-        self.interface = dbus.Interface(mng_proxy, _NMDbusManager.IF_NAME)
+    # libnm 1.12+ has this function
```
Are you saying we should keep the `_placeholder_ctx()`?

If not, since no one is checking its return, no need to do a return.

cathay4t replied:
Done.

cathay4t commented on libnmstate/netapplier.py:
```diff
@@ -82,10 +82,8 @@ def _verify_change(ifaces_desired_state):
-        checkpoint_ctx = nm.checkpoint.CheckPoint()
-    else:
-        checkpoint_ctx = _placeholder_ctx()
+    nm.checkpoint.has_checkpoint_capability()
```
Done.

cathay4t commented:
V2 patch uploaded.

Changes since V1:
 * Fixes based on feedback. Except https://github.com/nmstate/nmstate/pull/236#discussion_r249663592 which is pending decision.


The gobject warning is unrelated to this PR, a simple script like below will also trigger gobject warning:

```python
main_loop = GLib.MainLoop()
user_data = {}
nmclient.checkpoint_create([], 0, cp_flag, None, create_cb, user_data)
main_loop.run()
nmclient.checkpoint_rollback(user_data['cp'].get_path(), None, rollback_cb,
                             user_data)
main_loop.run()
```

If anything required fix this gobject warning, it shoule be in `nmclient.py` in another PR.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,115 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+    mainloop = checkpoint._mainloop
+    try:
+        cp = client.checkpoint_create_finish(result)
+        checkpoint._checkpoint = cp
```
You need the attribute to be public in order to access it.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,115 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            userdata)
+
+        if not self._mainloop.run(timeout):
+            raise CheckPointError("Failed to create checkpoint")
```
You better add the error info: `self._mainloop.error`

cathay4t replied:
Done

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,115 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                self._checkpoint.get_path(),
+                self._mainloop.cancellable,
+                _checkpoint_rollback_callback,
+                userdata)
```
nit: Based on how you started the open parenthesis, the closing one better be on its own line.

cathay4t replied:
Done

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,115 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                self._mainloop.cancellable,
+                _checkpoint_destroy_callback,
+                userdata)
+            self._checkpoint = None
```
Shouldn't you destroy it in `_checkpoint_destroy_callback`?
Same for the rollback.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,115 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            if not self._mainloop.run(self._timeout):
+                raise CheckPointError("Failed to destroy checkpoint")
+        else:
+            logging.warning('No checkpoint to destroy')
```
Shouldn't we raise?

cathay4t replied:
Change to raise `CheckPointError`

cathay4t commented:
@EdDev 

V3 patch uploaded.

Changes since V2:
 * Applied the suggestions in feedback.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,122 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+            iface = path
+        else:
+            iface = d.get_iface()
+        logging.debug('Interface %s rollback succeeded', iface)
```
This should be under the `else` of the following condition, right?
Perhaps use `if ret[path] == 0` if this is indeed the case.

cathay4t replied:
Done.

EdDev commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,122 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+    logging.debug('Checkpoint %s rollback executed', cp.get_path())
+    for path in ret:
+        d = client.get_device_by_path(path)
+        if d is None:
```
This can be one liner: `iface = path if d is None else d.get_iface()`

cathay4t replied:
Done.

cathay4t commented on libnmstate/netapplier.py:
```diff
@@ -31,6 +31,8 @@
 from libnmstate.prettystate import format_desired_current_state_diff
 from libnmstate.schema import Constants
 
+CHECKPOINT_MAINLOOP_TIMEOUT = 2
```
Done.

cathay4t commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,122 @@ def __exit__(self, exc_type, exc_val, exc_tb):
-        self._checkpoint = cp
+
+        if not self.mainloop.run(timeout):
+            raise CheckPointError("Failed to create checkpoint: "
```
Done for single quotes.

If I don't get your suggest wrong, I changed it to:

```python
        ok = self.mainloop.run(timeout)                                         
        if not ok:
```

cathay4t commented on libnmstate/nm/checkpoint.py:
```diff
@@ -92,21 +47,122 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                userdata
+            )
+            if not self.mainloop.run(self._timeout):
+                raise CheckPointError("Failed to destroy checkpoint: "
```
Done.

cathay4t commented:
@EdDev @Tyll Let's put a halt on this PR. A 'Segmentation fault' is triggered, waiting NM-dev to take a look. If that's problem, we need it been fixed in libnm(and updated in RHEL 7) before merge it.

cathay4t commented:
Please revisit this PR as RHEL 7.7 has the sigfault fix now.

New patch uploaded.

Changes:
 * Rebase to master.
 * Removed the `time.sleep(5)` in `netapplier.py` which is used for waiting rollback before disabling IPv6. Currently it's not needed.

EdDev commented:
> Please revisit this PR as RHEL 7.7 has the sigfault fix now.

But there is no upstream on which we can test this (except Fedora I guess).
What about the coming CentOS 8.0? Was this included in RHEL 8.0?

cathay4t commented:
> > Please revisit this PR as RHEL 7.7 has the sigfault fix now.
> 
> But there is no upstream on which we can test this (except Fedora I guess).
> What about the coming CentOS 8.0? Was this included in RHEL 8.0?

RHEL 8.0, Fedora 29/30 don't have that fix.
And CentOS 7.7 is not released yet.

Let's wait CentOS 8.1 or Fedora 31 release.

EdDev commented:
> RHEL 8.0, Fedora 29/30 don't have that fix.
> And CentOS 7.7 is not released yet.
> 
> Let's wait CentOS 8.1 or Fedora 31 release.

CentOS 7.7 is release already and we have CI we can test this on now.
Fedora 31 GA is planned in two weeks or so.

The only problem I see at the moment is that if we switch now, without checking the NM version, we will not be able to support Fedora 30.

tyll commented:
Fedora 31 is now released

cathay4t commented:
Wait on other PR #543

cathay4t commented:
New patch uploaded.

Changes:
 * Use `libnmstate.nm.nmclient.mainloop(refresh=True)` in checkpoint to make sure rollback does not impacted by previous mainloop failure.

tyll commented on libnmstate/nm/checkpoint.py:
```diff
@@ -17,23 +17,10 @@
 # along with this program. If not, see <https://www.gnu.org/licenses/>.
 #
 
+from contextlib import contextmanager
```
this import does not seem to be used according to LGT

tyll commented on tests/integration/nm/checkpoint.py:
```diff
@@ -93,3 +93,7 @@ def test_creating_a_checkpoint_from_dbuspath():
     assert not get_checkpoints()
+
+def test_repeat_create_destory_checkpoint():
+    for _ in range(0, 1000):
```
Why 1000? How long does this take?

cathay4t replied:
less than 1 second.

tyll commented on libnmstate/nm/checkpoint.py:
```diff
@@ -91,10 +43,11 @@ def get_checkpoints():
-        self._dbuspath = dbuspath
         self._autodestroy = autodestroy
+        self._dbuspath = dbuspath
+        self._mainloop = nmclient.mainloop(refresh=True)
```
IMHO it would be nicer to not use a different mainloop just for checkpoint management.

tyll commented on libnmstate/nm/checkpoint.py:
```diff
@@ -110,43 +63,158 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                        checkpoint._dbuspath
+                    )
+                )
+                mainloop.execute_next_action()
```
IMHO it would be nicer to embrace the libnm async API here and instead of using the checkpoint internal main loop to callback to a method from netapplier to continue with setting up the profiles once the checkpoint is created.

tyll commented on packaging/Dockerfile.fedora-nmstate-dev:
```diff
@@ -42,10 +40,7 @@ RUN dnf -y install --setopt=install_weak_deps=False \
-                   \
-                   # Below package for pip (used by tox) to build dbus-python
                    make \
-                   dbus-devel && \
```
the `&&` needs to be kept

tyll commented on packaging/Dockerfile.fedora-nmstate-dev:
```diff
@@ -16,8 +16,6 @@ RUN dnf -y install --setopt=install_weak_deps=False \
                    openvswitch \
```
the CentOS 8 container spec needs to be updated as well

cathay4t commented:
> The container spec for Fedora breaks with this PR. Is this an important change? It is not in the current sprint and a rather complex change.

Not important but good to have. This PR was on hold for our CI changes, as we bumped our CI to F31 and CentOS 8, let's try to merged it.
It's in current print: https://nmstate.atlassian.net/browse/NMSTATE-102

EdDev commented:
@cathay4t , there are still CI failures on this one.
We should try to take it in, but I think it has lower priority from the rest. If we cannot get it in for 0.1.1, then lest consider this a fix for 0.2.0.

tyll commented on tests/integration/nm/checkpoint.py:
```diff
@@ -93,3 +93,7 @@ def test_creating_a_checkpoint_from_dbuspath():
     new_checkpoint.destroy()
     assert not get_checkpoints()
+
+def test_repeat_create_destory_checkpoint():
```
There is a typo in the name, it should say `destroy`.
